crates/sui-core/src/consensus_throughput_calculator.rs

```diff
@@ -0,0 +1,517 @@
+            if current_throughput >= *threshold {
+                return *profile;
+            }
+        }
+
+        warn!("Could not resolve traffic profile for throughput {} - we shouldn't end up here. Fallback to lowest profile as default.", current_throughput);
+
+        // If not found, then we should return the lowest possible profile as default.
+        self.lowest_profile()
+    }
+}
+
+impl Default for TrafficProfileRanges {
+    fn default() -> Self {
+        let profiles = vec![
+            TrafficProfile {
+                level: Level::Low,
+                throughput: 0,
+            },
+            TrafficProfile {
+                level: Level::High,
+                throughput: 2_000,
+            },
+        ];
+        TrafficProfileRanges::new(&profiles)
+    }
+}
+
+pub type TimestampSecs = u64;
+
+#[derive(Debug)]
+pub struct TrafficProfileEntry {
+    /// The traffic profile
+    profile: TrafficProfile,
+    /// The time when this traffic profile was created
+    timestamp: TimestampSecs,
+    /// The calculated throughput when this profile created
+    #[allow(unused)]
+    throughput: u64,
+}
+
+#[derive(Default)]
+struct ConsensusThroughputCalculatorInner {
+    observations: VecDeque<(TimestampSecs, u64)>,
+    total_transactions: u64,
+}
+
+pub struct ConsensusThroughputCalculator {
+    /// The number of transaction traffic observations that should be stored within the observations
+    /// vector in the ConsensusThroughputCalculatorInner. Those observations will be used to calculate
+    /// the current transactions throughput. We want to select a number that give us enough observations
+    /// so we better calculate the throughput and protected against spikes. A large enough value though
+    /// will make us less reactive to traffic changes.
+    observations_window: u64,
+    /// The time that should be passed between two consecutive traffic profile updates. For example, if
+    /// we switch at point T to profile "Low", there will need to be passed at least `traffic_profile_update_window`
+    /// seconds until the traffic profile gets updated to a different value.
+    traffic_profile_update_window: TimestampSecs,
+    /// When current calculated throughput (A) is lower than previous, and the assessed profile is now a lower than previous,
+    /// we'll change to the lower profile only when (A) <= (previous_profile.throughput) * (100 - traffic_profile_cool_down_throughput_threshold) / 100.
+    /// Otherwise we'll stick to the previous profile. We want to do that to avoid any jittery behaviour that alternates between two profiles.
+    traffic_profile_cool_down_throughput_threshold: u64,
+    inner: Mutex<ConsensusThroughputCalculatorInner>,
+    last_traffic_profile: ArcSwap<TrafficProfileEntry>,
+    current_throughput: AtomicU64,
+    metrics: Arc<AuthorityMetrics>,
+    profile_ranges: TrafficProfileRanges,
+}
+
+impl ConsensusThroughputCalculator {
```

Comment:
My suggestion is to break this into two separate components:
- ThroughputGauge: behaves like a normal gauge. We can keep using the custom gauge implementation here.
- ThroughputProfiler: adjusts profile based on reading from the gauge.

This will be easier to test and reason about.

Reply:
Thanks for the suggestion @mwtian . I was actually very in between breaking down to multiple components versus doing all at one as the `ConsensusThroughputCalculator` . I went with one component approach primarily for two reasons
* currently there is no need to re-use/share the gauge separately from the profiler - as soon as there was that need we could split
* updating the profile is following the consensus output and is triggered by it - using a profiler will defer us to a pull based approach which would depend on the time of pulling/updating the info. Now that might not be _that_ important and maybe the alternative it would be for the gauge to call a list of listeners (ex the profiler) to pass them the updates. 

let me know your thoughts. But I would be keen on doing this refactoring if you think that it would help here.

Reply:
> behaves like a normal gauge. We can keep using the custom gauge implementation here.

what do you mean by normal gauge? Are you referring to an existing component we use in our codebase? I chose to go with this custom one to use the timestamps coming from consensus to ensure things are as consistent as possible across validators.